### PR TITLE
HDDS-8690. Ozone Support deletion related parameter dynamic configuration

### PR DESCRIPTION
**Pull request by xichen01**
## What changes were proposed in this pull request?
Support deletion related parameter dynamic configuration
OM:
- ozone.key.deleting.limit.per.task

SCM:
- hdds.scm.block.deletion.per-interval.max

Datanode:
- hdds.datanode.block.deleting.limit.per.interval
- hdds.datanode.block.delete.threads.max
- ozone.block.deleting.service.workers

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8690

## How was this patch tested?

unit test.

### CODE REVIEW

**Review by adoroszlai — COMMENTED** [review]
Thanks @xichen01 for working on this.  Mostly looks good, but I would prefer to implement this after #4774, #4794 (which both need some committer reviews), and a follow-up that connects the annotated configurations with the reconfiguration handler.  The change will be much simpler then.

**Comment by xichen01** [comment]
> Thanks @xichen01 for working on this. Mostly looks good, but I would prefer to implement this after #4774, #4794 (which both need some committer reviews), and a follow-up that connects the annotated configurations with the reconfiguration handler. The change will be much simpler then.

OK, I will update this PR after https://github.com/apache/ozone/pull/4774 and https://github.com/apache/ozone/pull/4794 were merged

**Comment by xichen01** [comment]
@adoroszlai Should we maintain `Reconfigurable properties lists` to describe which keys can be configured, this may require effort to keep the documentation in sync with the code. 
Perhaps we could remove this section, and the user can get exactly this information via the command?
https://github.com/apache/ozone/blob/9c6cd4b93e0427fc12bbedae8156c746cbe6cea1/hadoop-hdds/docs/content/feature/Reconfigurability.md?plain=1#L41-L48

**Comment by adoroszlai** [comment]
> Should we maintain `Reconfigurable properties lists` to describe which keys can be configured, this may require effort to keep the documentation in sync with the code. Perhaps we could remove this section, and the user can get exactly this information via the command?

Yes, let's remove the specific properties.  Documenting the way to explore the list of properties is enough.  An outdated doc may be worse than no doc.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java [hunk 1]
```diff
@@ -2125,6 +2128,14 @@ private String reconfOzoneReadOnlyAdmins(String newVal) {
     return String.valueOf(newVal);
   }
 
+  private String reconfHddsScmBlockDeletionPerIntervalMax(String newVal) {
+    getConfiguration().set(HDDS_SCM_BLOCK_DELETION_PER_INTERVAL_MAX, newVal);
+
+    getScmBlockManager().getSCMBlockDeletingService()
+        .setBlockDeleteTXNum(Integer.parseInt(newVal));
+    return newVal;
+  }
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java:2137]: We simplify reconfiguration of this property by using `reconfigurable = true` and making sure the same instance of `scmConfig` is used in `SCMBlockDeletingService` as the one being reconfigured in SCM.

Full change (also fixes `TestScmReconfiguration` failure):
https://github.com/adoroszlai/hadoop-ozone/commit/1ce698fab5f9b6a55dab6e28f382bb29f6d65cc0

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/reconfig/TestDatanodeReconfiguration.java [hunk 2]
```diff
@@ -18,21 +18,94 @@
...
+        OZONE_BLOCK_DELETING_SERVICE_WORKERS,
+        String.valueOf(originPoolSize + 1));
+    assertEquals(originPoolSize + 1, executor.getCorePoolSize());
+
+    // Attempt to decrease the pool size by 1 and verify if it's successful
+    getFirstDatanode().getReconfigurationHandler().reconfigurePropertyImpl(
+        OZONE_BLOCK_DELETING_SERVICE_WORKERS,
+        String.valueOf(originPoolSize - 1));
+    assertEquals(originPoolSize - 1, executor.getCorePoolSize());
+  }
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/reconfig/TestDatanodeReconfiguration.java:105]: We can avoid duplication by converting to `@ParameterizedTest`: https://github.com/adoroszlai/hadoop-ozone/commit/5aeb768560b36bf182dcc96bc50a0c71625f279d

Also fixes findbugs error:

```
M D RV: Return value of org.apache.hadoop.ozone.container.keyvalue.statemachine.background.BlockDeletingService.getBlockLimitPerInterval() ignored, but method has no side effect  At TestDatanodeReconfiguration.java:[line 58]
```

https://github.com/xichen01/ozone/actions/runs/5577408467/jobs/10190274215#step:6:2367

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/reconfig/TestOmReconfiguration.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
 import static org.apache.commons.lang3.RandomStringUtils.randomAlphabetic;
 import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_ADMINISTRATORS;
 import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_READONLY_ADMINISTRATORS;
+import static org.apache.hadoop.ozone.om.OMConfigKeys.OZONE_KEY_DELETING_LIMIT_PER_TASK;
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/reconfig/TestOmReconfiguration.java:29]: `reconfigurableProperties()` test case needs to be updated: https://github.com/adoroszlai/hadoop-ozone/commit/9d9226e0afe713e70418874d26a156d2ea862688

Fixes https://github.com/xichen01/ozone/actions/runs/5577408467/jobs/10190272872#step:5:4011

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java [hunk 4]
```diff
@@ -669,4 +668,12 @@ public long getBytesReleased() {
       return bytesReleased;
     }
   }
+
+  public int getBlockLimitPerInterval() {
+    return dfsConf.getBlockDeletionLimit();
+  }
+
+  public DatanodeConfiguration getDfsConf() {
+    return dfsConf;
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java:677]: Nit: I think `dnConf` / `getDnConf` would be better.

**Review by adoroszlai — COMMENTED** [review]
Thanks @xichen01 for continuing work on this.  Mostly looks good.

**Review by adoroszlai — COMMENTED** [review]
Thanks @xichen01 for updating the patch.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/conf/OzoneConfiguration.java [hunk 5]
```diff
@@ -335,4 +336,16 @@ private static void addDeprecatedKeys() {
...
+  @Override
+  public synchronized <T> T getSingletonObject(Class<T> configurationClass) {
+    if (singletons.containsKey(configurationClass)) {
+      return (T) singletons.get(configurationClass);
+    }
+
+    T singletonObject = getObject(configurationClass);
+    singletons.put(configurationClass, singletonObject);
+    return singletonObject;
+  }
```
[the remark this answers is not in the document]
Comment (xichen01) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/conf/OzoneConfiguration.java]: Done.

File: hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationSource.java [hunk 6]
```diff
@@ -183,6 +183,20 @@ default <T> T getObject(Class<T> configurationClass) {
...
+   * stored in a map for future retrieval.
+   *
+   * @param configurationClass The class for which a singleton
+   *                           instance is required
+   * @return a singleton instance of the given class
+   */
+  default <T> T getSingletonObject(Class<T> configurationClass) {
+    return getObject(configurationClass);
+  }
+
```
[the remark this answers is not in the document]
Comment (xichen01) [reply hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationSource.java]: `getSingletonObject` has been removed from `ConfigurationSource.java`

[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationSource.java]: Thanks @xichen01 for making the change.  On second thought, instead of adding a new method `getSingletonObject`, we should override `getObject` from `ConfigurationSource` with the singleton-based implementation.  This way we can keep passing `ConfigurationSource`.

In the long run, we should move towards passing the type-safe config objects (`DatanodeConfiguration`, `ScmConfig`, etc.)

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/conf/OzoneConfiguration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/conf/OzoneConfiguration.java]: It doesn't need to be `synchronized`, since it uses a `ConcurrentHashMap`'s atomic operation.

File: hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationSource.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (xichen01) [reply hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationSource.java]: If we change the logic of `ConfigurationSource#getObject` directly, will this cause some of the tests to fail, since at the moment we can assume that each object from `ConfigurationSource#getObject` is independent and unaffected.
If the `ConfigurationSource#getObject` is "Singleton" that maybe cause tests to affect with each other.

Should we keep this implemented(`getSingletonObject`) temporarily, and implement a "Singleton" `ConfigurationSource#getObject` in another MR?

**Comment by ChenSammi** [comment]
@xichen01 , could you rebase this patch once you got time?

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/conf/OzoneConfiguration.java [hunk 7]
```diff
@@ -335,4 +334,5 @@ private static void addDeprecatedKeys() {
             OZONE_CONTAINER_COPY_WORKDIR)
     });
   }
+
```
Comment (ChenSammi) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/conf/OzoneConfiguration.java:337]: Can you revert the changes in this file since it's actually doesn't change any content?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java [hunk 8]
```diff
@@ -260,7 +261,7 @@ public DatanodeStateMachine(DatanodeDetails datanodeDetails,
 
   @VisibleForTesting
   public DatanodeStateMachine(DatanodeDetails datanodeDetails,
-                              ConfigurationSource conf) throws IOException {
+                              OzoneConfiguration conf) throws IOException {
```
Comment (ChenSammi) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java:264]: Why we change the conf from ConfigurationSource to OzoneConfiguration? it's also changed in below OzoneContainer.java.

File: hadoop-hdds/docs/content/feature/Reconfigurability.md [hunk 9]
```diff
@@ -38,13 +38,18 @@ The meaning of command options:
     - **status**:     Check reconfig status
     - **properties**: List reconfigurable properties
 
-## OM Reconfigurability
+## Retrieve the reconfigurable properties list
+To retrieve all the reconfigurable properties list for a specific component in Ozone,
+you can use the command: `ozone admin reconfig properties --address=<ip:port>`.
```
Comment (ChenSammi) [inline hadoop-hdds/docs/content/feature/Reconfigurability.md:43]: `ozone admin reconfig properties --address=<ip:port>` doesn't work.  The accepted format is what you used in the below examples.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 10]
```diff
@@ -4704,6 +4707,14 @@ private String reconfOzoneReadOnlyAdmins(String newVal) {
     return String.valueOf(newVal);
   }
 
+  private String reconfOzoneKeyDeletingLimitPerTask(String newVal) {
+    getConfiguration().set(OZONE_KEY_DELETING_LIMIT_PER_TASK, newVal);
```
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:4711]: Please add the check that newValue should be a positive value.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java [hunk 11]
```diff
@@ -137,7 +146,7 @@ public class DatanodeConfiguration {
    * missed. With max threads 5, optimistically DN can handle 1500 individual
    * container delete tx in 60s with RocksDB cache miss.
    */
-  @Config(key = "block.delete.threads.max",
+  @Config(key = BLOCK_DELETE_THREAD_MAX,
```
[the remark this answers is not in the document]
Comment (xichen01) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:149]: If we change a configuration's value, if only need to change the value itself, and no additional operations are performed.  we can just add a `reconfigurable = true` in `@Config` annotation. the reconfigurable configuration can be auto registered. 
```java
      reconfigurationHandler =
          new ReconfigurationHandler("DN", conf, this::checkAdminPrivilege)
              .register(HDDS_DATANODE_BLOCK_DELETE_THREAD_MAX,
                  this::reconfigBlockDeleteThreadMax)
              .register(OZONE_BLOCK_DELETING_SERVICE_WORKERS,
                  this::reconfigDeletingServiceWorkers)
              .register(conf.getObject(DatanodeConfiguration.class));  // register the reconfigurable configuration
```
then this  configuration's value can be modify in runtime by `ozone admin reconfig` command.

But If need to do something extra when modifying the configuration value, such as modifying the thread pool (`block.delete.threads.max`), we need to register the operation manually.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 10 above)
Comment (xichen01) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:4713]: Done.

Comment (xichen01) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:4713]: I think may we can provide a `conf.getUnsignedInt()` for those configurations which must be larger than 0. we can do some check in `conf.getUnsignedInt()`

File: hadoop-hdds/docs/content/feature/Reconfigurability.md (hunk 9 above)
Comment (xichen01) [reply hadoop-hdds/docs/content/feature/Reconfigurability.md]: Already corrected

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/conf/OzoneConfiguration.java (hunk 7 above)
Comment (xichen01) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/conf/OzoneConfiguration.java]: Done.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:165]: So either register the property explicitly with a consumer handler, or set the  "reconfigurable = true"  attribute to the property.  The later way the property will be auto registered with a default consumer handler, right? @xichen01

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 10 above)
Comment (ChenSammi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:4720]: You mean conf.setUnsignedInt?   It's a good idea. But the configuration class is in hadoop common module.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java (hunk 8 above)
Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java:269]: @xichen01 , is this one missed?

**Comment by ChenSammi** [comment]
@xichen01 , thanks for the improving the patch. There is one comment missed to address. And could you check the failed unit test?  I'm not sure if it's related or not with the code change of this patch.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (xichen01) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:165]: Yes, if you set the "reconfigurable = true" attribute, the configuration will be auto handled when "reconfigure".
but one point needs to notice:

The object registered to `ReconfigurationHandler` must be the same conf as the one that wants to get the value after reconfiguring.
such as:
```java

      DatanodeConfiguration dnConf1 = conf.getObject(DatanodeConfiguration.class)
      DatanodeConfiguration dnConf2 = conf.getObject(DatanodeConfiguration.class)

      reconfigurationHandler =
          new ReconfigurationHandler("DN", conf, this::checkAdminPrivilege)
              .register(dnConf1); 
```
After reconfiguring, only `dnConf1 ` can get the changed value, because only `dnConf1` was registered to the `ReconfigurationHandler`

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java (hunk 8 above)
Comment (xichen01) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java]: Done

**Comment by ChenSammi** [comment]
@xichen01 , could you rebase the patch to solve the conflicts again?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:165]: @xichen01 , thanks for the explain.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java [hunk 12]
```diff
@@ -668,4 +679,24 @@ private void checkAdminPrivilege(String operation)
   public ReconfigurationHandler getReconfigurationHandler() {
     return reconfigurationHandler;
   }
+
+  private String reconfigBlockDeleteThreadMax(String value) {
+    getConf().set(HDDS_DATANODE_BLOCK_DELETE_THREAD_MAX, value);
+    conf.getObject(DatanodeConfiguration.class)
+        .setBlockDeleteThreads(Integer.parseInt(value));
```
Comment (ChenSammi) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java:686]: Looks like this line is covered by L684 already.  It can be removed.

**Comment by ChenSammi** [comment]
@xichen01 , can you check the failed test "TestOzoneManagerHAKeyDeletion.testKeyDeletion"? It might be related with the code change.

**Comment by xichen01** [comment]
> @xichen01 , can you check the failed test "TestOzoneManagerHAKeyDeletion.testKeyDeletion"? It might be related with the code change.

The master branch have this Error too https://github.com/apache/ozone/actions/runs/6439086990/job/17486681800

**Comment by ChenSammi** [comment]
> > @xichen01 , can you check the failed test "TestOzoneManagerHAKeyDeletion.testKeyDeletion"? It might be related with the code change.
> 
> The master branch have this Error too https://github.com/apache/ozone/actions/runs/6439086990/job/17486681800

Yes, it's reproduced with master branch.  Created HDDS-9414 to track the failure.

**Comment by xichen01** [comment]
> > > @xichen01 , can you check the failed test "TestOzoneManagerHAKeyDeletion.testKeyDeletion"? It might be related with the code change.
> > 
> > 
> > The master branch have this Error too https://github.com/apache/ozone/actions/runs/6439086990/job/17486681800
> 
> Yes, it's reproduced with master branch. Created [HDDS-9414](https://issues.apache.org/jira/browse/HDDS-9414) to track the failure.

@adoroszlai has create a ticket
https://issues.apache.org/jira/projects/HDDS/issues/HDDS-9412

**Review by ChenSammi — APPROVED** [review]
